### PR TITLE
make var string return var char w/ BackwardsIndex

### LINKED ISSUE #14497: proc `[]`(s: string; i: BackwardsIndex): char {...} doesn't return var char
### Summary
Like 
proc `[]`[T](s: var openArray[T]; i: BackwardsIndex): var T {...}
proc `[]`[Idx, T](a: var array[Idx, T]; i: BackwardsIndex): var T {...}

can 
proc `[]`(s: string; i: BackwardsIndex): char {...} return var char?

### Description
I was trying to get this to work doing exercisim.io

```
func reverse*(a: string):string =
  result = a
  for i in 0 ..< a.len div 2:
    #swap(result[i], result[^(i+1)]) # result[BackwardsIndex(i+1)] is immutable, not var!
    swap(result[i], result[a.high - i]) # had to do this
```

### PR DESCRIPTION
**Pull request by metagn**
fixes #14497

### CODE REVIEW

**Comment by metagn** [comment]
There is a VM bug here that is very weird:

```nim
const foo = "foo"

proc proc1(s: var string) =
  if s[^1] notin {'a'..'z'}:
    s = ""

proc proc2(f: string): string =
  result = f
  proc1(result)

const bar = proc2(foo) # Error: attempt to access a nil address kind: rkInt
```

I will just work around it in this PR and submit it as a separate issue.

Edit: There are bigger bugs with `var T` in the VM with way more scope than I can comprehend. Reporting this specific incident doesn't seem to be very helpful because if you try to make it simpler a bigger bug appears. For now I've just reported #15463.

File: lib/system.nim [hunk 1]
```diff
@@ -2578,6 +2578,7 @@ proc `[]`*[T](s: var openArray[T]; i: BackwardsIndex): var T {.inline.} =
   system.`[]`(s, s.len - int(i))
 proc `[]`*[Idx, T](a: var array[Idx, T]; i: BackwardsIndex): var T {.inline.} =
   a[Idx(a.len - int(i) + int low(a))]
+proc `[]`*(s: var string; i: BackwardsIndex): var char {.inline.} = s[s.len - int(i)]
```
Comment (Araq) [inline lib/system.nim:2581]: Note: This addition means we'll merge it after 1.4.0

Comment (xflywind) [reply lib/system.nim:2581]: Any news?

**Comment by ringabout** [comment]
needs a changelog and removes workaround? Though it is the bug of VM, but this PR maybe cause problems without intention.

**Comment by metagn** [comment]
I would actually rather the VM bug be fixed and we don't use the workaround, but there's no guarantee that the VM bug will be fixed if no one notices it, so maybe this should stay open either way. Not my call though, it's a 1 line change + test, takes minimal effort to do

**Comment by ringabout** [comment]
@hlaaftana 
Since this PR is merged, maybe open an issue for the bug of VM?

**Comment by metagn** [comment]
The example I gave above works now, presumably with https://github.com/xflywind/Nim/commit/a67ac4a7ad4a566138f009599f83591f3f544aec, and the following example works as well with the `pathnorm` code using a nightly from before this PR was accepted:

```nim
import os
proc `[]`*(s: var string; i: BackwardsIndex): var char {.inline.} = s[s.len - int(i)]
const a = "a" / "b"
```

so this PR might have actually not needed the workaround (sorry for not realizing this if it's true), but #15464 still fails, in which case I can't think of another bug that could come up here.

**Comment by ringabout** [comment]
Maybe make a PR removing workaround or add a testcase for this.

**Comment by metagn** [comment]
Would it be fine if you or someone else did it? I am not equipped for any similar problems that may come up with the PR that removes the workaround, since you fixed the related bug you would be more prepared to figure out any problem.

**Comment by ringabout** [comment]
All right, I will try to make a PR.

**Comment by zedeus** [comment]
Jester triggers the VM bug, failing to compile now. 
Example:
```nim
import jester

routes:
  get "/?":
    resp Http200
```

**Pull request merged**